### PR TITLE
chore: impl custom directive for JIT

### PR DESCRIPTION
**Pull request by laststylebender14**
**Summary:**  
_Briefly describe the changes made in this PR._

**Issue Reference(s):**  
Fixes #... _(Replace "..." with the issue number)_

**Build & Testing:**

- [x] I ran `cargo test` successfully.
- [x] I have run `./lint.sh --mode=fix` to fix all linting issues raised by `./lint.sh --mode=check`.

**Checklist:**

- [x] I have added relevant unit & integration tests.
- [ ] I have updated the [documentation] accordingly.
- [x] I have performed a self-review of my code.
- [x] PR follows the naming convention of `<type>(<optional scope>): <title>`

[documentation]: https://github.com/tailcallhq/tailcallhq.github.io/tree/develop/docs

### CODE REVIEW

File: src/core/ir/resolver_context_like.rs [hunk 1]
```diff
@@ -84,17 +84,30 @@ impl From<async_graphql::SelectionField<'_>> for SelectionField {
...
         Self::from_jit_field(value)
     }
 }
 
 impl SelectionField {
-    fn from_jit_field<Input: ToString>(
-        field: &crate::core::jit::Field<Nested<Input>, Input>,
+    fn from_jit_field(
+        field: &crate::core::jit::Field<
+            Nested<async_graphql_value::ConstValue>,
```
Comment (meskill) [inline src/core/ir/resolver_context_like.rs:108]: maybe just use import, all that `async_graphql_value::` looks redundant

File: src/core/jit/model.rs [hunk 2]
```diff
@@ -314,3 +319,58 @@ impl<Input> OperationPlan<Input> {
...
+    pub fn try_map<Output, Error>(
+        self,
+        map: impl Fn(Input) -> Result<Output, Error>,
+    ) -> Result<Directive<Output>, Error> {
+        Ok(Directive {
+            name: self.name,
+            arguments: self
+                .arguments
+                .into_iter()
+                .filter_map(|(k, v)| map(v).ok().map(|v| (k, v)))
```
Comment (meskill) [inline src/core/jit/model.rs:372]: why do you ignore any possible errors here making the call infallible?

File: src/core/jit/model.rs [hunk 3]
```diff
@@ -314,3 +319,58 @@ impl<Input> OperationPlan<Input> {
         self.flat.len()
     }
 }
+
+/// DirectiveAdapter helps to bridge the gap between async_graphql's
+/// ConstDirective and our custom Directive implementation.
+pub trait DirectiveAdapter {
+    fn name(&self) -> &str;
+    fn arguments(&self) -> Vec<(String, &ConstValue)>;
```
Comment (meskill) [inline src/core/jit/model.rs:327]: seems inconsistent to have owned String and reference. Could we return both references and evade string cloning?

Comment (meskill) [reply src/core/jit/model.rs:327]: Also, returning Iterator could be preferred since it's not involving allocation for the vec if not needed

**Review by meskill — DISMISSED** [review]

File: src/core/document.rs [hunk 4]
```diff
@@ -61,7 +63,9 @@ fn get_formatted_docs(docs: Option<String>, indent: usize) -> String {
     formatted_docs
 }
 
-pub fn print_directives<'a>(directives: impl Iterator<Item = &'a ConstDirective>) -> String {
+pub fn print_directives<'a, Input: DirectiveAdapter + 'a>(
+    directives: impl Iterator<Item = &'a Input>,
+) -> String {
```
Comment (tusharmath) [inline src/core/document.rs:68]: Little unclear to me why this should change?

File: src/core/jit/model.rs [hunk 5]
```diff
@@ -360,3 +379,54 @@ impl<Input> OperationPlan<Input> {
...
+    fn name(&self) -> &str {
+        &self.name.node
+    }
+
+    fn arguments(&self) -> impl Iterator<Item = (&str, &ConstValue)> {
+        self.arguments
+            .iter()
+            .map(|(k, v)| (k.node.as_str(), &v.node))
+    }
+}
```
Comment (tusharmath) [inline src/core/jit/model.rs:400]: We can create a `DirectiveAdapter::from(const_directive)` instead which performs a ZERO copy transformation.

**Review by tusharmath — APPROVED** [review]

**Review by tusharmath — APPROVED** [review]

**Pull request merged**